### DISCUSSION TITLE
Is it possible to connect a real robot (as a device client) to an IoT hub?

### DISCUSSION THREAD

**Discussion by addy1997** [Azure/azure-iot-sdk-c#2580] [General]
Hi there,

I am working on a project to send data from robotic arm to my IoT hub. I understand that the robot in this case behaves as device client that needs to connect to the IoT hub. How do I send data from the robot to the hub?

Do I need to write a code (assuming the robot as a device client similar to the [pnp thermostat](https://github.com/Azure/azure-iot-sdk-c/tree/main/iothub_client/samples/pnp))?

**Comment by ewertons** [comment]
Hi @addy1997 ,
thanks for sharing about your project.
You can perfectly send data from your robot to the Azure IoT Hub.
Writing some code will be necessary, but to better guide you, it's always good to share details about the platform you are using (which microcontroller or CPU and operating system, compilers or development IDE).
Could you share those details? 
Thanks,
Azure IoT SDK Team.

**Comment by addy1997** [comment]
Hi @ewertons,

Thanks for your reply. So, I want to send the data from [Panda robotic arm](https://www.wevolver.com/specs/panda) to the IoT hub. I am using Ubuntu 20.04 OS and Visual Studio code for compilation. Azure iot Explorer and CLI are already installed on my device and I used Azure-iot-sdk to send telemetry data from my device client to the IoT hub.

Here are some queries I have:

1. Assuming the robot as a device client, how do I develop a production-ready code to send data from the robot to the hub?
2. Can I use the [pnp sample](https://github.com/Azure/azure-iot-sdk-c/tree/main/iothub_client/samples/pnp) to develop the code for the robot?

Do you need additional information?

**Comment by addy1997** [comment]
Hey @ewertons,

Could you help?

**Comment by ewertons** [comment]
Hi @addy1997 ,

Assuming the robot as a device client, how do I develop a production-ready code to send data from the robot to the hub?
> That is **a very broad** question. What is your specific question? production-ready code means you have tested your project over the platform it is intended to run on, you have tests for every error your dependencies might return, any changes in the system, etc. There are lot of considerations for your question. Could you be more specific?

Can I use the [pnp sample](https://github.com/Azure/azure-iot-sdk-c/tree/main/iothub_client/samples/pnp) to develop the code for the robot?
> Yes you can. It depends on your plans for your product. Taking dependency on PnP adds dependency on the capabilities, complexity and limitations of PnP (https://learn.microsoft.com/en-us/azure/iot/concepts-convention ). If you have a very small and well defined set of functionalities you want to use in Azure IoT Hub, I would recommend you simplify and do not use PnP.

**Comment by addy1997** [comment]
Hi @ewertons,  

That is a very broad question. What is your specific question? production-ready code means you have tested your project over the platform it is intended to run on, you have tests for every error your dependencies might return, any changes in the system, etc. There are lot of considerations for your question. Could you be more specific?

> > 
> > Broadly speaking, I plan to construct a digital twin of the [panda robotic arm](https://ori.ox.ac.uk/robots/panda-arm/) using DTDL and use IoT azure SDK for data transmission between the digital twin and the robotic arm. The objective is to control the movements of the robotic arm using the digital twin. 
> > Currently, I am able to send telemetry from the robot to IoT hub but I am not sure about how to use the telemetry with the digital twin.
> What are the considerations you have about my question?

**Comment by ericwolz** [comment]
Digital twin is online/offline state storage for devices.
Telemetry is use to provide insights in real time for [analytics](https://learn.microsoft.com/en-us/azure/stream-analytics/stream-analytics-introduction), enrichment, transformation,  and routed to storage.

This is already very well documented on the Internet.
https://learn.microsoft.com/en-us/azure/iot/iot-introduction
https://learn.microsoft.com/en-us/azure/iot/iot-overview-message-processing

**Comment by addy1997** [comment]
Hey @ewertons,

I am able to send telemetry from Device(my Robot) to the IoT hub and from Cloud to my device. Can this telemetry be passed on to the digital twin of the robot? (See [this](https://learn.microsoft.com/en-us/azure/digital-twins/how-to-ingest-iot-hub-data#example-telemetry-scenario))

**Reply by ewertons** [reply]
Yes. The link you provided already gives an example of how to do it with Azure Functions.
